OKTA-438760 Convert Third-party risk integration to single page guide (#2872)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/third-party-risk-integration/main/index.md

---
title: Third-party risk provider integration
excerpt: Provides documentation on configuring an Okta org to receive risk events from a third-party provider
layout: Guides
---

This guide shows you how to configure an Okta org to receive risk events from a third-party provider.

---
<!-- Nutrition facts bullets -->
**Learning outcomes**

* Learn how to set up and test third-party risk events with a sample app from the Postman Risk API collection.

**What you need**

* [Okta Developer Edition organization](https://developer.okta.com/signup)
* A configured Okta org for third-party risk providers. Contact your Okta customer support representative for this configuration.
* A [Postman client](https://www.postman.com/downloads/) to configure and test the risk provider integration. See [Get Started with the Okta APIs](/code/rest/) to set up Postman.
* The Risk API collection. Download from: [![Run in Postman](https://run.pstmn.io/button.svg)](https://app.getpostman.com/run-collection/47546754d382762468c6)

---

<ApiLifecycle access="ea" />

> The Okta Risk Engine evaluates authentication attempts by reviewing the risk score of the sign-in based on context and historical data. Using Okta Risk APIs, third-party risk providers can integrate with the Okta Risk Engine using a standard Okta service application. The third-party risk provider can send risk events, which can be used when calculating the authentication risk based on the risk policy configured in the Okta org. The risk events are additionally logged as part of the System Log.

Reviewer: ```suggestion
The Okta Risk Engine evaluates authentication attempts by reviewing the risk score of the sign-in based on context and historical data. Using Okta Risk APIs, third-party risk providers can integrate with the Okta Risk Engine using a standard Okta service application. The third-party risk provider can send risk events that you can use when calculating the authentication risk based on the risk policy configured in the Okta org. The risk events are additionally logged as part of the System Log.
```

> >**Note:** Third-party risk events are shared with and received from Non-Okta Applications.  Non-Okta Applications include web-based, offline, mobile, or other software application functionality that are provided by you or a third party and interoperate with the Okta Service. You are not required to receive or utilize third-party risk events within Okta Risk Engine, but if you configure Okta Risk Engine to utilize third-party risk events, then you agree on behalf of your organization that Okta may receive and share data with the Non-Okta Application as necessary to provide this functionality. You may only utilize these third-party risk events if you are a customer of both Okta and the Non-Okta Application. Okta cannot guarantee continued partnerships or functionality with any Non-Okta Applications.

Reviewer: ```suggestion
>**Note:** Third-party risk events are shared with and received from non-Okta applications. Non-Okta applications include web-based, offline, mobile, or other software application functionality that are provided by you or a third party and interoperate with the Okta Service. You aren't required to receive or use third-party risk events within Okta Risk Engine, but if you configure Okta Risk Engine to use third-party risk events, then you agree on behalf of your organization that Okta may receive and share data with the non-Okta application as necessary to provide this functionality. You may only use these third-party risk events if you are a customer of both Okta and the non-Okta application. Okta can't guarantee continued partnerships or functionality with any non-Okta applications.
```

Creating a third-party risk provider integration follows the general configurations for creating an OAuth service application using the OAuth client credentials grant flow. The service application provides an integration for the default risk provider and the Okta Risk Engine, and Risk Event API calls can test for a successful setup. Follow the high-level steps below to set up an example third-party risk provider integration.

1. [Create a self-service application for the risk provider](#create-a-service-app-for-a-third-party-risk-provider)
2. [Update the default risk provider](#update-the-default-third-party-risk-provider)
3. [Test the integration](#test-the-integration)

## Create a service app for a third-party risk provider

Your Okta org requires the set up of an OAuth service application to integrate and consume risk events from a third-party risk provider.

Use the following high-level steps to configure this service application:

1. [Create a public/private key pair](#create-a-public-private-key-pair)
2. [Create and verify the service application](#create-a-service-application)
3. [Add a scope grant to the service application](#add-a-scope-grant-to-the-application)

> For information on this general process, see [Implement OAuth for Okta with a Service App](/docs/guides/implement-oauth-for-okta-serviceapp/).

Reviewer: ```suggestion
See [Implement OAuth for Okta with a Service App](/docs/guides/implement-oauth-for-okta-serviceapp/).
```

    }
    ```

4. Copy the **Public Key** JSON value to your Postman environment's `publicKey` variable.
5. Make sure to save the JSON value for the **Public and Private Keypair**, which is required during testing.

For background information on this process, see [Create a public/private key pair](/docs/guides/implement-oauth-for-okta-serviceapp/main/#create-a-public-private-key-pair).

### Create a service application

Create the service application that integrates with the third-party risk provider using the previously generated public key for authentication.

1. Copy the name of your third-party risk provider to your Postman environment's `providerName` variable. In this example, use `Risk Provider Example`. Save the environment.

2. Call the following POST API from the Risk Integration Postman collection: **Admin: API to create OAuth service client (for the provider)** (`https://${yourOktaDomain}/oauth2/v1/clients`).

> 3. Review the response, which includes the `jwks` key pair. Note, the `client_name` value is the same value you gave as the `providerName` variable. A sample response follows:

Reviewer: ```suggestion
3. Review the response, which includes the `jwks` key pair. The `client_name` value is the same value that you gave as the `providerName` variable. A sample response follows:
```

    "client_id": "0oaaaboyxsbrWdsk81d6",
    "client_id_issued_at": 1611263018,
    "client_name": "Risk Provider Example",
    "client_uri": null,
    "logo_uri": null,
    "redirect_uris": [
        "https://httpbin.org/get"
    ],
    "response_types": [
        "token"
    ],
    "grant_types": [
        "client_credentials"
    ],
    "jwks": {
        "keys": [
            {
                "kty": "RSA",
                "alg": "RS256",
                "kid": null,
                "use": "sig",
                "e": "AQAB",
                "n": "uNj5u6PSWYR0VOTEhVsVEMfiFCVcNBKyRGM0YHQjGlhEHO-28Dw68l8U1KdHdiNVrvL21S-bfQFyQWSTF5_w5x966SmNMPHtjkxoJ_BOxyed3bKkbfLZgq8GM5lsAwTE-NIbMmPciX9Z4VEapaiKbEqg3KSGVzJcEH18E8AiIMgQ0ts7NTJ33sOwtqdTsQfho5crqtPIy1Z0Svvraq-UA7-elDWj9duqLE-YIRx-6U9hdBJ5Q7bC12H_TwcyNoLLwvtdi2X8NNV93CLJ1xoDAS9o8FDmWWUqciXq4XLww1kJRFvOMMT7LLefYdhV1Ef7MTZrpTOlwoDfDDdltTUUmw"
            }
        ]
    },
    "token_endpoint_auth_method": "private_key_jwt",
    "application_type": "service"
    }
    ```

4. From the response, copy the `client_id` value, in this example `0oaaaboyxsbrWdsk81d6`, to your Postman environment's `clientId` variable. Save the environment.

> 5. Verify the application is available by calling the following GET API from the Risk Integration Postman collection: **Admin: API to get all OAuth service clients** (`https://${yourOktaDomain}/oauth2/v1/clients`).

Reviewer: ```suggestion
5. Verify that the application is available by making a GET request using the following API from the Risk Integration Postman collection: **Admin: API to get all OAuth service clients** (`https://${yourOktaDomain}/oauth2/v1/clients`).
```

This call retrieves all service applications from your Okta org, including the new Risk Integration service application.

For background information on this process, see [Create a service app and grant scopes](/docs/guides/implement-oauth-for-okta-serviceapp/main/#create-a-service-app-and-grant-scopes/).

### Add a scope grant to the application

You must now define the allowed Risk scope for use with the third-party risk provider service application.

> 1. Call the following POST API from the Risk Integration Postman collection: **Admin: API to grant scopes to the OAuth service client** (`/api/v1/apps/${clientId}/grants`).

Reviewer: ```suggestion
1. Make a POST request using the following API from the Risk Integration Postman collection: **Admin: API to grant scopes to the OAuth service client** (`/api/v1/apps/${clientId}/grants`).
```

>     This call adds the `scopeId` to  the value of: `okta.riskEvents.manage`.

Reviewer: ```suggestion
    This call adds the `scopeId` to the value of: `okta.riskEvents.manage`.
```

2. Review the response for this addition. A sample portion of the response follows:

    ```JSON
     "_embedded": {
        "scope": {
            "id": "okta.riskEvents.manage"
        }
    },
    ```

For background information on granting scopes, see [Grant allowed scopes](/docs/guides/implement-oauth-for-okta-serviceapp/main/#grant-allowed-scopes).

> >**Note:** If you get a "Resource Not Found" exception in this step, it's possible that your `clientId` or the URL are not valid.

Reviewer: ```suggestion
>**Note:** If you get a "Resource Not Found" exception in this step, it's possible that your `clientId` or the URL aren't valid.
```

## Update the default third-party risk provider

> The Okta org contains a default risk provider profile, which must be configured for the third-party risk provider by your Okta administrator using the risk provider API. On your Okta org, you can have as many as three third-party risk providers available to send risk events to the Okta Risk Engine.

Reviewer: ```suggestion
The Okta org contains a default risk provider profile that must be configured for the third-party risk provider by your Okta administrator using the risk provider API. In your Okta org, you can have as many as three third-party risk providers available to send risk events to the Okta Risk Engine.
```

> Each third-party risk provider can also be configured to send three action types to the Okta Risk Engine: `none`, `log_only` (default), and `enforce_and_log`. In this example, the action is set to `enforce_and_log`, which uses the third-party risk event when calculating the risk for a sign-in.

Reviewer: ```suggestion
You can also configure each third-party risk provider to send three action types to the Okta Risk Engine: `none`, `log_only` (default), and `enforce_and_log`. In this example, the action is set to `enforce_and_log`, which uses the third-party risk event when calculating the risk for a sign-in.
```

> For further detail on the Risk Provider API see the following reference documentation: [Risk Provider API](/docs/reference/api/risk-providers).

Reviewer: ```suggestion
See the following reference documentation: [Risk Provider API](/docs/reference/api/risk-providers).
```

In this example, use the following two procedures to set up your third-party risk provider:
- [Retrieve the default risk provider](#retrieve-the-default-risk-provider)
- [Update the risk provider ](#update-the-risk-provider)

### Retrieve the default risk provider

This procedure retrieves the default risk provider profile and Provider ID.

> 1. Call the following GET API from the Risk Integration Postman collection: **Admin: API to get all Provider Settings** (`https://${yourOktaDomain}/api/v1/risk/providers`).

Reviewer: ```suggestion
1. Make a GET request using the following API from the Risk Integration Postman collection: **Admin: API to get all Provider Settings** (`https://${yourOktaDomain}/api/v1/risk/providers`).
```

> 2. Review the response, which includes the Provider ID, default name, and action properties. A sample response follows:

Reviewer: ```suggestion
2. Review the response that includes the Provider ID, default name, and action properties. A sample response follows:
```

    ```JSON
    [
    {
        "id": "rkpa9y5jpfe8l4ktr1d6",
        "name": "Default ",
        "clientId": "null",
        "action": "log_only",
        "_links": {
            "self": {
                "href": "https://test.oktapreview.com/api/v1/risk/providers/rkpa9y5jpfe8l4ktr1d6",
                "hints": {
                    "allow": [
                        "GET",
                        "PUT",
                        "DELETE"
                    ]
                }
            }
        }
    }
    ]
    ```

3. From the response, copy the `id` value, in this example `rkpa9y5jpfe8l4ktr1d6`, to your Postman environment's `providerId` variable.

### Update the risk provider

This procedure updates the default risk provider profile with the service application ID, risk provider name, and the risk provider action.

> 1. Call the following PUT API from the Risk Integration Postman collection: **Admin: API to Update Provider Settings** (`https://${yourOktaDomain}/api/v1/risk/providers/${providerId}`), which updates the default risk provider's data. This data is included in the request body:

Reviewer: ```suggestion
1. Make a PUT request using the following API from the Risk Integration Postman collection: **Admin: API to Update Provider Settings** (`https://${yourOktaDomain}/api/v1/risk/providers/${providerId}`). This call updates the default risk provider's data. The following data is included in the request body:
```

    ```JSON
    {
    "name": "{{providerName}}",
    "clientId": "{{clientId}}",
    "action": "enforce_and_log"
    }
    ```

> 2. Review the response from the call to make sure the risk provider now contains the required data. A sample response follows:

Reviewer: ```suggestion
2. Review the response from the call to make sure that the risk provider now contains the required data. A sample response follows:
```

    ```JSON
    {
    "id": "rkpa9y5jpfe8l4ktr1d6",
    "name": "Risk Provider Example",
    "clientId": "0oaaaboyxsbrWdsk81d6",
    "action": "enforce_and_log",
    "_links": {
        "self": {
            "href": "https://test.oktapreview.com/api/v1/risk/providers/rkpa9y5jpfe8l4ktr1d6",
            "hints": {
                "allow": [
                    "GET",
                    "PUT",
                    "DELETE"
                ]
            }
        }
        }
    }
    ```

## Test the integration

With the creation of a service application for the third-party risk provider, and the update of the third-party risk provider profile, you can now test the integration using the Risk Events API.

In practice, this API is used by the third-party provider to send risk events to an Okta org for risk policy evaluation or the system log. The third-party provider requires access to the public/private key pair and the service application ID (`clientId`) created by the Okta administrator in the [Create service application](#create-a-service-app-for-a-third-party-risk-provider) section.

> In this test, the API sends a sample payload risk event to the Okta org, which can be consumed by Okta and used to calculate the risk of the authentication.

Reviewer: ```suggestion
In this test, the API sends a sample payload risk event to the Okta org that Okta can then consume and use to calculate the risk of the authentication.
```

>**Note:** Only IP-related risk events are available for consumption by the Okta org.

Use the following high-level steps to test the risk provider integration:

1. [Create a client assertion for the access token](#create-a-client-assertion)
2. [Create an access token](#create-an-access-token)
3. [Send a risk event to the Okta org](#send-a-risk-event-to-okta)
4. [Confirm the response and system log](#confirm-the-response)

### Create a client assertion

> This procedure creates a signed JSON Web Token (JWT), which is used as the client assertion value required in the request for a scoped access token.

Reviewer: ```suggestion
This procedure creates a signed JSON Web Token (JWT) that is used as the client assertion value required in the request for a scoped access token.
```

1. Navigate to [Generate JWT](https://www.jsonwebtoken.dev/) to create a JWT.

> 2. In the **JWK KEY** field, copy the **Public and Private Keypair** generated when you created the service application ([Create a public/private key pair](#create-a-public-private-key-pair)).

Reviewer: ```suggestion
2. In the **JWK KEY** field, copy the **Public and Private Keypair** that you generate you create the service application ([Create a public/private key pair](#create-a-public-private-key-pair)).
```

> 3. In the **Payload** field, add the following JSON payload, substituting your service application ID (client ID) and your Okta org URL:

Reviewer: ```suggestion
3. In the **Payload** field, add the following JSON payload and substitute your service application ID (client ID) and your Okta org URL:
```

    ```JSON
    {
    "aud": "https://<url>/oauth2/v1/token",
    "iss": "<clientId>",
    "sub": "<clientId>"
    }
    ```

4. Click **Generate JWT**.
5. Copy the resulting JWT value to your Postman's `clientAssertion` environment variable. Save the environment.

For further background information on this process, see [Create and sign the JWT](/docs/guides/implement-oauth-for-okta-serviceapp/main/#create-and-sign-the-jwt/).

### Create an access token

This procedure creates an access token using the `clientAssertion` value required for authentication into the risk provider service application.

> 1. Call the following POST API from the Risk Integration Postman collection: **Partner: API to get the access token** (`https://${yourOktaDomain}/oauth2/v1/token`).

Reviewer: ```suggestion
1. Make a POST request using the following API from the Risk Integration Postman collection: **Partner: API to get the access token** (`https://${yourOktaDomain}/oauth2/v1/token`).
```

> 2. Review the response from the call, and copy the `access_token` value to your Postman's `accessToken` environment variable. Don't include the leading and trailing double quotes around the `access_token` value, below, while saving the `accessToken` variable. A sample response follows:

Reviewer: ```suggestion
2. Review the response from the call and copy the `access_token` value to your Postman's `accessToken` environment variable. Don't include the leading and trailing double quotes around the `access_token` value, below, while saving the `accessToken` variable. A sample response follows:
```

    "scope": "okta.riskEvents.manage"
    }
    ```
    >**Note**: The access token expires in 60 minutes (or the value set in the property `expires_in`).

For further background information on this process, see [Get an access token](/docs/guides/implement-oauth-for-okta-serviceapp/main/#get-an-access-token).

### Send a risk event to Okta

This procedure sends a sample risk event payload to the Okta org.

> 1. Call the following POST API from the Risk Integration Postman collection: **Partner: API to send RiskEvents (Auth using access token)** (`/api/v1/risk/events/ip`). A sample payload follows, which includes two events:

Reviewer: ```suggestion
1. Make a POST request using the following API from the Risk Integration Postman collection: **Partner: API to send RiskEvents (Auth using access token)** (`/api/v1/risk/events/ip`). A sample payload follows, which includes two events:
```

    ```JSON
    [
     {
       "timestamp": "2021-01-20T00:00:00.001Z",
       "subjects": [
          {
            "ip": "6.7.6.7",
            "riskLevel": "LOW",
            "message": "none"
          },
          {
            "ip": "1.1.1.1",
            "riskLevel": "HIGH" ,
            "message": "Detected Attack tooling and suspicious activity"
          }
        ]
     }
    ]
    ```

2. Review the request status of the API call. If the status is `202 Accepted`, the risk events were consumed by the Okta org.

> **Note:** Rate limits of three calls per minute, per risk provider, apply to the `/api/v1/risk/events/ip` endpoint. Each call can contain multiple risk events.

> For reference information on this API, see [Risk Events](/docs/reference/api/risk-events).

Reviewer: ```suggestion
See [Risk Events](/docs/reference/api/risk-events).
```

### Confirm the response

This procedure reviews the Admin Console's System Log to identify the risk event information logged after calling the Risk Event API.

1. Sign in to your Okta org as an administrator.
2. In the Admin Console, go to **Reports** > **System Log**.

> 3. Review the log file or search for the event `security.risk.signal.consume`, which is logged when a risk provider sends a risk event to Okta.

Reviewer: ```suggestion
3. Review the log file or search for the event `security.risk.signal.consume` that is logged when a risk provider sends a risk event to Okta.
```

> 4. With a risk action of `enforce_and_log`, and a risk-based policy setup, the third-party risk provider event is used when calculating the authentication risk. This information is logged in the `user.session.start` event.

Reviewer: ```suggestion
4. With a risk action of `enforce_and_log`, and a risk-based policy set up, the third-party risk provider event is used when calculating the authentication risk. This information is logged in the `user.session.start` event.
```

### General troubleshooting tips
* Save the Postman environment after every change to the environment. Confirm that you are using the correct Postman environment.

> * Postman environment variables are case sensitive. Ensure there are no typos, no leading/trailing spaces, no leading or trailing double quotes in the environment variables.

Reviewer: ```suggestion
* Postman environment variables are case sensitive. Ensure that there are no typos, no leading/trailing spaces, no leading or trailing double quotes in the environment variables.
```

> * The org URL should not be the admin URL (for example, use `https://demo-org.oktapreview.com` instead of  `https://demo-org-admin.oktapreview.com`) and you don't need a trailing `/` at the end of the URL.

Reviewer: ```suggestion
* The org URL shouldn't be the admin URL (for example, use `https://demo-org.oktapreview.com` instead of  `https://demo-org-admin.oktapreview.com`). You don't need a trailing `/` at the end of the URL.
```
